expression/expression.go

```diff
@@ -95,7 +113,49 @@ func (triggerExpression *TriggerExpression) Evaluate() (moira.State, error) {
 	}
 }
 
-func validateUserExpression(triggerExpression *TriggerExpression, userExpression *govaluate.EvaluableExpression) (*govaluate.EvaluableExpression, error) {
+// Validate gets trigger expression and validates it for given parameters using expr
+func (triggerExpression *TriggerExpression) Validate() error {
+	expression, err := getExpression(triggerExpression)
```

Comment:
Right now your solution is doing validation even for triggers from simple mode, maybe you should make it do it only for user expressions?
I suggest this change: you can leave the `triggerExpression.Evaluate()` call, but change the `validateUserExpression` function

Then in essence there will be 2 steps of running the expression:
- in `expr` the compilation stage
- in `govaluate` the stage of program execution

Reply:
Okay, I'll make the following changes:
- remove the Validate() function
- the current actions in the Validate() function should be in the `validateUserExpression` function after checking for variables.
- this has to still be cached because we don't want to validate an expression that has already been validated before because checker calls `Evaluate` frequently.

Look at this and let me know if this is good to go @almostinf

Reply:
Yeah, it's a good plan:
- Move all your logic to the `validateUserExpression` function, thus not breaking the previous code, but adding only a new validation step
- Also, you won't need to change the previous tests much, `Evaluate()` will be called in them as well, you will just need to add more test cases
- Also, if `expr.Run()`, checks all arguments for validity, you don't need to validate them a second time

Reply:
This will however alter the expected error for some of the tests

Reply:
I don't think it's that scary